main.nf

```diff
@@ -469,83 +520,252 @@ if (params.peptide_fasta){
+      file(umis_per_cell)
+      set val(channel_id), file(good_barcodes) into good_barcodes_ch
+
+      script:
+      umis_per_cell = "${channel_id}__n_umi_per_cell.csv"
+      good_barcodes = "${channel_id}__barcodes.tsv"
+      """
+      count_umis_per_cell.py \\
+          --reads ${reads} \\
+          --min-umi-per-cell ${tenx_min_umi_per_cell} \\
+          --cell-barcode-pattern '${tenx_cell_barcode_pattern}' \\
+          --molecular-barcode-pattern '${tenx_molecular_barcode_pattern}' \\
+          --umis-per-cell ${umis_per_cell} \\
+          --good-barcodes ${good_barcodes}
+      """
+    }
+    // Make sure good barcodes file is nonempty so next step doesn't start
+    // it[0] = channel id
+    // it[1] = good_barcodes file
+    // good_barcodes_unfiltered_ch.filter{ it[1].size() > 0 }
+    //   .ifEmpty{ log.info "No cell barcodes found with at least ${tenx_min_umi_per_cell} molecular barcodes (UMIs) per cell"}
+    //   .set{ good_barcodes_ch }
+
+  } else {
+    // Use barcodes extracted from the tenx .tgz file
+    good_barcodes_ch = tenx_bam_barcodes_ch
+  }
+  tenx_reads_ch.cross(good_barcodes_ch)
+    .map{ it -> tuple(it[0][0], it[0][1], it[0][2], it[1][1]) }
+    .dump(tag: "tenx_reads_with_counts_ch" )
+    .set{ tenx_reads_with_good_barcodes_ch }
+
+  process extract_per_cell_fastqs {
+    tag "${channel_id}_${is_aligned}"
+    publishDir "${params.outdir}/10x-fastqs/per-cell/${channel_id}/", mode: 'copy'
+
+    input:
+    // Example output:
+    // [[ANTOINE_TESTIS, 'aligned', ANTOINE_TESTIS__aligned.fastq.gz],
+    //   [ANTOINE_TESTIS, ANTOINE_TESTIS__barcodes.tsv]]
+    set val(channel_id), val(is_aligned), file(reads), file(barcodes) from tenx_reads_with_good_barcodes_ch
+
+    output:
+    file('*.fastq.gz') into per_channel_cell_reads_ch
+
+    script:
     """
-      bam2fasta convert \\
-        $processes \\
-        $min_umi_per_barcode \\
-        $line_count \\
-        $rename_10x_barcodes \\
-        $barcodes_file \\
-        $save_fastas \\
-        $save_intermediate_files \\
-        $metadata \\
-        --filename $bam
-      find . -type f -name "*.fasta" | while read src; do if [[ \$src == *"|"* ]]; then mv "\$src" \$(echo "\$src" | tr "|" "_"); fi done
+    make_per_cell_fastqs.py \\
+        --reads ${reads} \\
+        --good-barcodes ${barcodes} \\
+        --cell-barcode-pattern '${tenx_cell_barcode_pattern}' \\
+        --channel-id ${channel_id}__${is_aligned}
     """
   }
+  // Make per-cell fastqs into a flat channel that matches the read channels of yore
+  per_channel_cell_reads_ch
+    .dump(tag: 'per_channel_cell_reads_ch')
+    .flatten()
+    .map{ it -> tuple(it.simpleName, file(it))}
+    .dump(tag: 'per_cell_fastqs_ch')
+    .set{ per_cell_fastqs_ch }
+
+
+  if (params.skip_trimming) {
+    reads_ch = ch_non_bam_reads.concat(per_cell_fastqs_ch)
+  } else {
+    ch_read_files_trimming = ch_non_bam_reads.concat(per_cell_fastqs_ch)
+  }
+}
+
+
+
+if (!params.skip_trimming){
```

Comment:
is this condition correct?

reads_ch is empty is the error and I am wondering if it arises from here 

it line 673 you declare reads_ch params.skip_trimming is true, but you are trying to use reads_ch when its the opposite